django/contrib/admindocs/utils.py

```diff
@@ -151,90 +154,188 @@ def replace_metacharacters(pattern):
+            end:    Ending index of the `group` inside the `pattern`.
+
+        Returns:
+            `tuple` containing the start and end indices of the named group.
+        """
+        nonlocal pattern
         unmatched_open_brackets, prev_char = 1, None
         for idx, val in enumerate(pattern[end:]):
-            # Check for unescaped `(` and `)`. They mark the start and end of a
-            # nested group.
             if val == '(' and prev_char != '\\':
                 unmatched_open_brackets += 1
             elif val == ')' and prev_char != '\\':
                 unmatched_open_brackets -= 1
             prev_char = val
-            # If brackets are balanced, the end of the string for the current
-            # named capture group pattern has been reached.
+
             if unmatched_open_brackets == 0:
-                group_pattern_and_name.append((pattern[start:end + idx + 1], group_name))
-                break
+                return start, (end + idx + 1)
+
+    group_indices = []
+    if named is True:
+        for start, end, group_name in unnamed_group_indices:
+            group_indices.append((*get_group_start_and_end_indices(start, end), group_name))
+    else:
+        for start, end in unnamed_group_indices:
+            group_indices.append(get_group_start_and_end_indices(start, end))
 
-    # Replace the string for named capture groups with their group names.
+    group_start_end_indices = []
+    prev_end = None
+    if named is True:
+        for start, end, group_name in group_indices:
+            if prev_end and start > prev_end or not prev_end:
+                group_start_end_indices.append((start, end, group_name))
+            prev_end = end
+    else:
+        for start, end in group_indices:
+            if prev_end and start > prev_end or not prev_end:
+                group_start_end_indices.append((start, end))
+            prev_end = end
+    return group_start_end_indices
+
+
+def replace_named_groups(pattern):
+    r"""
+    Finds named groups in the given `pattern` and replaces them with the group
+    name.
+
+        * ^(?P<a>\w+)/b/(\w+)$ ==> ^<a>/b/(\w+)$
+        * ^(?P<a>\w+)/b/(?P<c>\w+)/$ ==> ^<a>/b/<c>/$
+        * ^(?P<a>\w+)/b/(\w+) ==> ^<a>/b/(\w+)
+        * ^(?P<a>\w+)/b/(?P<c>\w+) ==> ^<a>/b/<c>
+
+    Finds the string for named capture groups inside the given `pattern` and the
+    group name and replaces the named capture group string with its group name
+    inside the `pattern`.
+
+    Args:
+        pattern: `Regex` pattern for a `named` group.
+
+    Returns:
+        `pattern` string where named group sub-strings are all replaced with just
+        the group name.
+    """
+    group_start_end_indices_and_name = _find_groups(pattern, named_group_matcher, True)
+    group_pattern_and_name = [(pattern[start:end], group_name)
+                              for start, end, group_name in group_start_end_indices_and_name]
```

Comment:
Use hanging indentation as described in [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).